pkgs/development/libraries/libbsd/default.nix

```diff
@@ -35,6 +35,9 @@ stdenv.mkDerivation rec {
     url = "https://gitlab.freedesktop.org/libbsd/libbsd.git";
   };
 
+   # Fix undefined reference errors with version script under LLVM.
+  NIX_LDFLAGS = lib.optionalString (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") "--undefined-version";
```

Comment:
Is it possible to get around the mass rebuild by using:

```
stdenv.mkDerivation (rec {
  # ...
} // lib.optionalAttrs (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") {
  NIX_LDFLAGS = "--undefined-version"
})
```

Or is this too janky?

Reply:
Maybe, `optionalString` is a bit more clear but this is what we've done already quite a bit in this situation. Keeping the pattern the same would be preferable. The better solution is to have upstream fix it but some packages don't want to deal with this unique setup.

Reply:
well, env vars should be under the `env` attribute anyway which makes this easy.

```nix
env = lib.optionalAttrs (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") {
  NIX_LDFLAGS = "--undefined-version";
};
```
but that generates an infinite recursion accessing `stdenv.cc.bintools` IIRC as we tried to do this before but ended up just using either `makeFlags` or `configureFlags` which were already used in the derivation to avoid rebuilds. however, none of those tricks will work for this one if the infinite recursion is real.

Reply:
Yeah though wouldn't `env` still be present if we were to use `optionalAttrs`? That would likely still cause a rebuild.

Reply:
> Yeah though wouldn't `env` still be present if we were to use `optionalAttrs`? That would likely still cause a rebuild.

no. it either already exists or it doesn't get added when empty.

Reply:
can test using hello by adding an empty `env = {};` and then call `nix-build -A hello` and see that no rebuild happens.

Reply:
Yes, using `env` works. I added it.

Reply:
Nevermind, causes infinite recursion somewhere else.

Reply:
yeah, it doesn't seem to like `stdenv.cc...`

Reply:
Yeah, though the way it is like this seems to work.

Reply:
this seems to "work"

```nix
  # Fix undefined reference errors with version script under LLVM.
  env = lib.optionalAttrs (stdenv.hostPlatform.linker == "lld") {
    NIX_LDFLAGS = lib.optionalString (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") "--undefined-version";
  };
```

Reply:
Yeah, it seems we end up with two solutions that prevents rebuilds but looks gross. Or the simplest solution which looks clean but causes a ton of rebuilds.

Reply:
`NIX_LDFLAGS` should be `env.NIX_LDFLAGS`. i guess there was a push to make sure all env vars are under the `env` attribute. not sure if it is documented tho.

Reply:
using `configureFlags` doesn't seem to cause a rebuild either. so can do:

```nix
configureFlags = lib.optionals (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") [ "LDFLAGS=-Wl,--undefined-version" ];
```

Reply:
Alright, moved to `env`.

Reply:
^ That's the most elegant solution, probably :-)

Reply:
> using `configureFlags` doesn't seem to cause a rebuild either. so can do:
> 
> ```nix
> configureFlags = lib.optionals (stdenv.cc.bintools.isLLVM && lib.versionAtLeast stdenv.cc.bintools.version "17") [ "LDFLAGS=-Wl,--undefined-version" ];
> ```

Sweet, moved to that.

Reply:
yeah, looking at <details><summary>`nix derivation show -f '<nixpkgs>' libbsd`</summary>

```json
{
  "/nix/store/y8j0bsyifw08sqsdnrz3djk1rwdjqn4y-libbsd-0.11.8.drv": {
    "args": [
      "-e",
      "/nix/store/v6x3cs394jgqfbi0a42pam708flxaphh-default-builder.sh"
    ],
    "builder": "/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/bash",
    "env": {
      "__structuredAttrs": "",
      "buildInputs": "",
      "builder": "/nix/store/306znyj77fv49kwnkpxmb0j2znqpa8bj-bash-5.2p26/bin/bash",
      "cmakeFlags": "",
      "configureFlags": "",
      "depsBuildBuild": "",
      "depsBuildBuildPropagated": "",
      "depsBuildTarget": "",
      "depsBuildTargetPropagated": "",
      "depsHostHost": "",
      "depsHostHostPropagated": "",
      "depsTargetTarget": "",
      "depsTargetTargetPropagated": "",
      "dev": "/nix/store/hrv0rl13qlmqic54mhzv98wllqvj6zkk-libbsd-0.11.8-dev",
      "doCheck": "1",
      "doInstallCheck": "",
      "enableParallelBuilding": "1",
      "enableParallelChecking": "1",
      "enableParallelInstalling": "1",
      "man": "/nix/store/jz2nrn8gncqcwgg9f1f9lrfjlbxxqhbr-libbsd-0.11.8-man",
      "mesonFlags": "",
      "name": "libbsd-0.11.8",
      "nativeBuildInputs": "/nix/store/2bx07vad8gfbs8kdsx0ylxhla715ay7x-autoreconf-hook",
      "out": "/nix/store/p222av5bbv60iq8pya7bh7rx0rx9gaxx-libbsd-0.11.8",
      "outputs": "out dev man",
      "patches": "",
      "pname": "libbsd",
      "propagatedBuildInputs": "/nix/store/2bvbxxq8vzq6lis903p8zmkngsxppmkv-libmd-1.1.0",
      "propagatedNativeBuildInputs": "",
      "src": "/nix/store/593j9841lbyrb5m0iggdxr7h2bys21dj-libbsd-0.11.8.tar.xz",
      "stdenv": "/nix/store/xfhkjnpqjwlf6hlk1ysmq3aaq80f3bjj-stdenv-linux",
      "strictDeps": "",
      "system": "x86_64-linux",
      "version": "0.11.8"
    },
    "inputDrvs": {
      "/nix/store/5nj0j17h95qfrv1h36y9h6inln166hz4-libmd-1.1.0.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/h37ch3ni4p03imz19nwnqijhr65s8vlq-autoreconf-hook.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/hj2f0byx1ikcbp13j6i2bnba6lpxfbca-bash-5.2p26.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/m0a1cy1jfbcqss7x4l8jwx8zh54p3k91-libbsd-0.11.8.tar.xz.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      },
      "/nix/store/panbw08d4yffvkwkl7fpa2k051sjwpz0-stdenv-linux.drv": {
        "dynamicOutputs": {},
        "outputs": [
          "out"
        ]
      }
    },
    "inputSrcs": [
      "/nix/store/v6x3cs394jgqfbi0a42pam708flxaphh-default-builder.sh"
    ],
    "name": "libbsd-0.11.8",
    "outputs": {
      "dev": {
        "path": "/nix/store/hrv0rl13qlmqic54mhzv98wllqvj6zkk-libbsd-0.11.8-dev"
      },
      "man": {
        "path": "/nix/store/jz2nrn8gncqcwgg9f1f9lrfjlbxxqhbr-libbsd-0.11.8-man"
      },
      "out": {
        "path": "/nix/store/p222av5bbv60iq8pya7bh7rx0rx9gaxx-libbsd-0.11.8"
      }
    },
    "system": "x86_64-linux"
  }
}
```
</details>


 shows that {cmake,configure,meson}Flags exist and are empty strings -- also shows they are all under `env` which i guess is why adding an empty `env = {};` doesn't cause a rebuild.

BTW the change is still using `env.LDFLAGS` and not `configureFlags`

Reply:
Seems I didn't push it, though I did.